"address" warnings about possible memory leaks due to lru_cache

I have mentioned while looking at https://github.com/datalad/datalad-fuse/runs/6672682413?check_suite_focus=true of #72 that linting gives us

```
lint run-test: commands[0] | flake8 --config=tox.ini datalad_fuse
datalad_fuse/fsspec.py:53:6: B019 Use of `functools.lru_cache` or `functools.cache` on methods can lead to memory leaks. The cache may retain instance references, preventing garbage collection.
datalad_fuse/fsspec.py:206:6: B019 Use of `functools.lru_cache` or `functools.cache` on methods can lead to memory leaks. The cache may retain instance references, preventing garbage collection.
datalad_fuse/fuse_.py:110:6: B019 Use of `functools.lru_cache` or `functools.cache` on methods can lead to memory leaks. The cache may retain instance references, preventing garbage collection.
```

Needs analysis on either it is something to just ignore or we indeed would retain too many instances.

--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -1,7 +1,7 @@
 exclude: ^(_datalad_buildsupport/.*|datalad_fuse/_version\.py|versioneer\.py)$
 repos:
   - repo: https://github.com/pre-commit/pre-commit-hooks
-    rev: v4.0.1
+    rev: v4.3.0
     hooks:
       - id: check-added-large-files
       - id: check-json
@@ -11,12 +11,12 @@ repos:
       - id: trailing-whitespace
 
   - repo: https://github.com/psf/black
-    rev: 21.9b0
+    rev: 22.3.0
     hooks:
       - id: black
 
   - repo: https://github.com/PyCQA/isort
-    rev: 5.9.3
+    rev: 5.10.1
     hooks:
       - id: isort
 

--- a/datalad_fuse/fsspec.py
+++ b/datalad_fuse/fsspec.py
@@ -1,6 +1,5 @@
 from datetime import datetime, timezone
 from enum import Enum
-from functools import lru_cache
 import logging
 import os
 import os.path
@@ -13,6 +12,7 @@ from datalad.utils import get_dataset_root
 import fsspec
 from fsspec.exceptions import BlocksizeMismatchError
 from fsspec.implementations.cached import CachingFileSystem
+import methodtools
 
 from .consts import CACHE_SIZE
 from .utils import is_annex_dir_or_key
@@ -50,7 +50,7 @@ class DatasetAdapter:
         if self.annex is not None:
             self.annex._batched.clear()
 
-    @lru_cache(maxsize=CACHE_SIZE)
+    @methodtools.lru_cache(maxsize=CACHE_SIZE)
     def get_file_state(self, relpath: str) -> Tuple[FileState, Optional[str]]:
         p = self.path / relpath
         lgr.debug("get_file_state: %s", relpath)
@@ -203,7 +203,7 @@ class FsspecAdapter:
             ds.close()
         self.datasets.clear()
 
-    @lru_cache(maxsize=CACHE_SIZE)
+    @methodtools.lru_cache(maxsize=CACHE_SIZE)
     # TODO: optimize "caching" more since for all files under the same directory
     # they all would belong to the same dataset
     def get_dataset_path(self, path: Union[str, Path]) -> Path:

--- a/datalad_fuse/fuse_.py
+++ b/datalad_fuse/fuse_.py
@@ -2,7 +2,7 @@ from ctypes import CDLL, c_int, c_void_p
 from ctypes.util import find_library
 from datetime import datetime
 from errno import ENOENT, EROFS
-from functools import lru_cache, wraps
+from functools import wraps
 import io
 import logging
 import os
@@ -14,6 +14,7 @@ from threading import Lock
 from datalad import cfg
 from datalad.distribution.dataset import Dataset
 from fuse import FuseOSError, Operations
+import methodtools
 
 from .consts import CACHE_SIZE
 from .fsspec import FsspecAdapter
@@ -107,7 +108,7 @@ class DataLadFUSE(Operations):  # LoggingMixIn,
             )
         )
 
-    @lru_cache(maxsize=CACHE_SIZE)
+    @methodtools.lru_cache(maxsize=CACHE_SIZE)
     def getattr(self, path, fh=None):
         # TODO: support of unlocked files... but at what cost?
         lgr.debug("getattr(path=%r, fh=%r)", path, fh)
@@ -381,7 +382,7 @@ def file_getattr(f, timestamp: datetime):
     else:
         data["st_mode"] = stat.S_IFREG | 0o644
         data["st_size"] = info["size"]
-        data["st_blksize"] = 5 * 2 ** 20
+        data["st_blksize"] = 5 * 2**20
         data["st_nlink"] = 1
     data["st_atime"] = timestamp.timestamp()
     data["st_ctime"] = timestamp.timestamp()

--- a/setup.cfg
+++ b/setup.cfg
@@ -16,6 +16,7 @@ python_requires = >= 3.6
 install_requires =
     datalad >= 0.13.0
     fsspec[fuse,http] >= 2022.1.0
+    methodtools ~= 0.4.5
 packages = find:
 include_package_data = True
 

